File: client/golang/mpc/generic_parameters.go
/*
 * Copyright (c) 2000, 2023, trustbe and/or its affiliates. All rights reserved.
 * TRUSTBE PROPRIETARY/CONFIDENTIAL. Use is subject to license terms.
 *
 *
 */

package mpc

import (
	"context"
	"github.com/be-io/mesh/client/golang/codec"
	"github.com/be-io/mesh/client/golang/log"
	"github.com/be-io/mesh/client/golang/macro"
)

func init() {
	var _ macro.Parameters = new(GenericParameters)
}

type GenericParameters map[string]interface{}

func (that GenericParameters) PutAll(dict map[string]interface{}) {
	for key, value := range dict {
		that[key] = value
	}
}

func (that GenericParameters) GetKind() interface{} {
	return (*map[string]interface{})(nil)
}

func (that GenericParameters) GetArguments(ctx context.Context) []interface{} {
	var arguments []interface{}
	for key, value := range that {
		if key == "attachments" {
			continue
		}
		arguments = append(arguments, value)
	}
	return arguments
}

func (that GenericParameters) SetArguments(ctx context.Context, arguments ...interface{}) {

}

func (that GenericParameters) GetAttachments(ctx context.Context) map[string]string {
	attachments := that["attachments"]
	if nil == attachments {
		return nil
	}
	if attach, ok := attachments.(map[string]string); ok {
		return attach
	}
	cdc, ok := macro.Load(codec.ICodec).Get("json").(codec.Codec)
	if !ok {
		log.Error(ctx, "Parse attachments without codec named json exist")
		return nil
	}
	buff, err := cdc.Encode(attachments)
	if nil != err {
		log.Error(ctx, err.Error())
		return nil
	}
	var attach map[string]string
	if _, err = cdc.Decode(buff, &attach); nil != err {
		log.Error(ctx, err.Error())
		return nil
	}
	return attach
}

func (that GenericParameters) SetAttachments(ctx context.Context, attachments map[string]string) {
	that["attachments"] = attachments
}

func (that GenericParameters) Get(key string) interface{} {
	return that[key]
}

File: client/java/src/main/java/com/be/mesh/client/mpc/GenericParameters.java
/*
 * Copyright (c) 2000, 2023, trustbe and/or its affiliates. All rights reserved.
 * TRUSTBE PROPRIETARY/CONFIDENTIAL. Use is subject to license terms.
 *
 *
 */
package com.be.mesh.client.mpc;

import com.be.mesh.client.tool.FitMap;

import java.io.Serializable;
import java.lang.reflect.Type;
import java.util.*;

/**
 * @author coyzeng@gmail.com
 */
public class GenericParameters extends TreeMap<String, Object> implements Serializable, Parameters {

    private static final long serialVersionUID = -4338158729657717760L;
    private static final String ATTACHMENTS = "attachments";

    @Override
    public Map<String, Object> map() {
        return this;
    }

    @Override
    public Class<?> type() {
        return this.getClass();
    }

    @Override
    public Object[] arguments() {
        List<Object> args = new ArrayList<>(this.size());
        this.forEach((key, value) -> {
            if (ATTACHMENTS.equals(key)) {
                return;
            }
            args.add(value);
        });
        return args.toArray();
    }

    @Override
    public void arguments(Object[] arguments) {
        //
    }

    @Override
    public Map<Integer, Type> argumentTypes() {
        return new HashMap<>();
    }

    @SuppressWarnings("unchecked")
    @Override
    public Map<String, String> attachments() {
        Object attachments = this.computeIfAbsent(ATTACHMENTS, key -> new HashMap<>());
        if (attachments instanceof Map) {
            return new FitMap<>((Map<String, String>) attachments);
        }
        Codec codec = ServiceLoader.load(Codec.class).get(Codec.JSON);
        Map<String, String> des = codec.decode(codec.encode(attachments), Types.MapString);
        this.put(ATTACHMENTS, des);
        return new FitMap<>(des);
    }

    @Override
    public void attachments(Map<String, String> attachments) {
        this.put(ATTACHMENTS, attachments);
    }
}
